concurrency/MSNSession.java

```diff
@@ -0,0 +1,512 @@
+                        JiveGlobals.getIntProperty("plugin.gateway.msn.connectport", 1863));
+            }
+            catch (Exception e) {
+                Log.debug("MSN user is not able to log in: " + msnMessenger.getOwner().getEmail(), e);
+                setFailureStatus(ConnectionFailureReason.UNKNOWN);
+                sessionDisconnected("Unable to log in.");
+            }
+        }
+    }
+
+    /**
+     * @see net.sf.kraken.session.TransportSession#logOut()
+     */
+    @Override
+    public void logOut() {
+        cleanUp();
+        sessionDisconnectedNoReconnect(null);
+    }
+
+    /**
+     * @see net.sf.kraken.session.TransportSession#cleanUp()
+     */
+    @Override
+    public void cleanUp() {
+        if (msnMessenger != null) {
+            if (msnListener != null) {
+                ((BasicMessenger)msnMessenger).removeSessionListener(msnListener);
+                msnMessenger.removeContactListListener(msnListener);
+                msnMessenger.removeEmailListener(msnListener);
+                msnMessenger.removeMessageListener(msnListener);
+                msnMessenger.removeMessengerListener(msnListener);
+                msnMessenger.removeSwitchboardListener(msnListener);
+            }
+            try {
+                msnMessenger.getOwner().setPersonalMessage("");
+            }
+            catch (Exception e) {
+                // Ignore then.
+            }
+            msnMessenger.logout();
+        }
+    }
+
+    /**
+     * Retrieves the manager for this session.
+     *
+     * @return Messenger instance the session is associated with.
+     */
+    public MsnMessenger getManager() {
+        return msnMessenger;
+    }
+
+    /**
+     * Records information about a group on the user's contact list.
+     *
+     * @param msnGroup MSN group we are storing a copy of.
+     */
+    public void storeGroup(MsnGroup msnGroup) {
+        msnGroups.put(msnGroup.getGroupName(), msnGroup);
+    }
+
+    /**
+     * Removes information about a group from the user's contact list.
+     *
+     * @param msnGroup MSN group we are removing a copy of.
+     */
+    public void unstoreGroup(MsnGroup msnGroup) {
+        msnGroups.remove(msnGroup.getGroupName());
+    }
+
+    /**
+     * Records a member of a pending new group that will be added later.
+     *
+     * @param groupName Name of group to be stored.
+     * @param member Email address of member to be added.
+     */
+    public void storePendingGroup(String groupName, Email member) {
+        if (!msnPendingGroups.containsKey(groupName)) {
+            ArrayList<Email> newList = new ArrayList<Email>();
+            newList.add(member);
+            msnPendingGroups.put(groupName, newList);
+        }
+        else {
+            ArrayList<Email> list = msnPendingGroups.get(groupName);
+            list.add(member);
+            msnPendingGroups.put(groupName, list);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **215 and 223**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).